velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -301,39 +427,52 @@ class MinMaxByAggregate : public exec::Aggregate {
   template <typename MayUpdate>
   inline void updateValues(
       char* group,
-      T newValue,
-      U newComparisonValue,
-      bool isNull,
+      const DecodedVector& decodedValues,
+      const DecodedVector& decodedComparisons,
+      vector_size_t index,
+      bool isValueNull,
       MayUpdate mayUpdate) {
     clearNull(group);
-    if (mayUpdate(comparisonValue(group), newComparisonValue)) {
-      value(group) = newValue;
-      comparisonValue(group) = newComparisonValue;
-      valueIsNull(group) = isNull;
+    if (mayUpdate(comparisonValue(group), decodedComparisons, index)) {
+      valueIsNull(group) = isValueNull;
+      if (LIKELY(!isValueNull)) {
+        store<T, ValueAccumulatorType>(
+            value(group), decodedValues, index, allocator_);
+      }
+      store<U, ComparisonAccumulatorType>(
+          comparisonValue(group), decodedComparisons, index, allocator_);
     }
   }
 
-  inline T& value(char* group) {
-    return *reinterpret_cast<T*>(group + Aggregate::offset_);
+  inline ValueAccumulatorType* value(char* group) {
+    return reinterpret_cast<ValueAccumulatorType*>(group + Aggregate::offset_);
   }
 
-  inline U& comparisonValue(char* group) {
-    return *reinterpret_cast<U*>(group + Aggregate::offset_ + sizeof(T));
+  inline ComparisonAccumulatorType* comparisonValue(char* group) {
+    return reinterpret_cast<ComparisonAccumulatorType*>(
+        group + Aggregate::offset_ + sizeof(ValueAccumulatorType));
   }
 
   inline bool& valueIsNull(char* group) {
     return *reinterpret_cast<bool*>(
-        group + Aggregate::offset_ + sizeof(T) + sizeof(U));
+        group + Aggregate::offset_ + sizeof(ValueAccumulatorType) +
+        sizeof(ComparisonAccumulatorType));
   }
 
-  // Initial value will take the minimum and maximum values of the numerical
-  // limits.
+  /// Initial value will take the minimum and maximum values of the numerical
+  /// limits. This is only meaningful if the comparison value is numeric type.
   const U initialValue_;
+  DecodedVector decodedValueVector_;
+  DecodedVector decodedComparisonVector_;
+  DecodedVector decodedIntermediateResultVector_;
```

Comment:
nit: you may want to drop Vector_ suffix for readability: decodedValue_, decodedComparison_, decodedIntermediateResult_